Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp

```diff
@@ -145,15 +146,117 @@ s32 NWC24MakeUserID(u64* nwc24_id, u32 hollywood_id, u16 id_ctr, HardwareModel h
 }  // Anonymous namespace
 
 NetKDRequestDevice::NetKDRequestDevice(Kernel& ios, const std::string& device_name)
-    : Device(ios, device_name), config{ios.GetFS()}
+    : Device(ios, device_name), config{ios.GetFS()}, dlList{ios.GetFS()}
 {
+  m_work_queue.Reset([this](AsyncTask task) {
+    const IPCReply reply = task.handler();
+    {
+      std::lock_guard lg(m_async_reply_lock);
+      m_async_replies.emplace(AsyncReply{task.request, reply.return_value});
+    }
+  });
 }
 
 NetKDRequestDevice::~NetKDRequestDevice()
 {
   WiiSockMan::GetInstance().Clean();
 }
 
+void NetKDRequestDevice::Update()
+{
+  {
+    std::lock_guard lg(m_async_reply_lock);
+    while (!m_async_replies.empty())
+    {
+      const auto& reply = m_async_replies.front();
+      GetIOS()->EnqueueIPCReply(reply.request, reply.return_value);
+      m_async_replies.pop();
+    }
+  }
+}
+
+IPCReply NetKDRequestDevice::HandleNWC24DownloadNowEx(const IOCtlRequest& request)
+{
+  std::vector<u8> file_data;
+
+  u32 flags = Memory::Read_U32(request.buffer_in);
+  // Nintendo converts the entry ID between a u32 and u16
+  // several times, presumably for alignment purposes.
+  // We'll skip past buffer_in+4 and keep the entry index as a u16.
+  u16 entryIndex = Memory::Read_U16(request.buffer_in + 6);
+  u32 subtaskBitmask = Memory::Read_U32(request.buffer_in + 8);
+
+  INFO_LOG_FMT(IOS_WC24,
+               "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - flags: {}, index: {}, bitmask: {}",
+               flags, entryIndex, subtaskBitmask);
+
+  INFO_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - URL: {}",
+               dlList.GetDownloadURL(entryIndex));
+
+  INFO_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - Name: {}",
+               dlList.GetVFFContentName(entryIndex));
+
+  INFO_LOG_FMT(IOS_WC24, "NET_KD_REQ: IOCTL_NWC24_DOWNLOAD_NOW_EX - NI - Subtask: {}",
+               dlList.SubtaskID(entryIndex));
+
+  const Common::HttpRequest::Response response = m_http.Get(dlList.GetDownloadURL(entryIndex));
```

Comment:
`dlList.GetDownloadURL(entryIndex)` can be defined in a variable.  That'll also make it easier to check if there are errors.